### ISSUE TITLE
Mangled contours if startpoint is offcurve

### ISSUE THREAD

**Issue by frankrolf** [roboDocs/ScaleFast#5]
I noticed that an interpolation with scaleFast may mangle curves if the starting point is inconsistent across masters – one contour may start with an offcurve, the other with an oncurve. While this doesn’t cause any problems in many interpolation scenarios, scaleFast isn’t that forgiving:

![Screen Shot 2019-09-26 at 12 11 10](https://user-images.githubusercontent.com/2049645/65760101-a11ee080-e11c-11e9-9d7f-253f50af610a.png)

I solved the issue by pre-processing my UFOs with this little script first, which makes sure all contours start with oncurve points:

```
from ufoLib.pointPen import PointToSegmentPen


def redraw_glyph(g):
    contours = []

    for contour in g.contours:
        points = [p for p in contour.points]
        while points[0].type == 'offcurve':
            points.append(points.pop(0))
        contours.append(points)

    g.prepareUndo('redraw')
    rg = RGlyph()
    ppen = PointToSegmentPen(rg.getPen())

    for contour in contours:
        ppen.beginPath()
        for point in contour:
            if point.type == 'offcurve':
                ptype = None
            else:
                ptype = point.type
            ppen.addPoint((point.x, point.y), ptype)
        ppen.endPath()

    g.clearContours()
    g.appendGlyph(rg)
    g.performUndo()

f = CurrentFont()
glyphs_with_contours = [g for g in f if len(g.contours)]
glyphs_with_offcurve_start = []

for g in glyphs_with_contours:
    for contour in g.contours:
        first_point = contour.points[0]
        first_bPoint = contour.bPoints[0]
        first_point_coords = (first_point.x, first_point.y)
        if first_point_coords != first_bPoint.anchor:
            glyphs_with_offcurve_start.append(g)

for g in glyphs_with_offcurve_start:
    print(g.name)
    redraw_glyph(g)

```

**Comment by frankrolf** [comment]
I prepared an example for this. Consider the two attached UFOs, which contain the same single glyph, one starting with oncurve, the other with offcurve.
Here is a script to interpolate the two:
```
import os
f_all = sorted([f for f in AllFonts() if f.path.endswith('curve.ufo')], key=lambda f: f.path)
f_off, f_on = f_all
g_off = f_off['a']
g_on = f_on['a']

inter_result = g_off + (g_off - g_on) * 0.5
f_target = RFont()
f_target.glyphOrder = ['a']
ng = f_target.newGlyph('a', clear=True)
ng.appendGlyph(inter_result)
ng.width = inter_result.width
f_target.save(os.path.expanduser('~/Desktop/target.ufo'))

```

As expected, the interpolation result looks like this:
![Screen Shot 2020-06-02 at 14 43 23](https://user-images.githubusercontent.com/2049645/83521578-b801a800-a4df-11ea-8df0-a467de9560ac.png)

However, ScaleFast does that:
![Screen Shot 2020-06-02 at 14 32 13](https://user-images.githubusercontent.com/2049645/83521599-c059e300-a4df-11ea-9792-0d0c4398b31d.png)

Might this simply be an issue of interpolating points vs interpolating bPoints?

[oncurve vs offcurve start.zip](https://github.com/roboDocs/ScaleFast/files/4717197/oncurve.vs.offcurve.start.zip)

**Comment by justvanrossum** [comment]
> As expected

I don't agree with this. I think it only _accidentally_ works for some reason (perhaps the same reason as why RF doesn't show the startpoint correctly in the offcurve ufo?). These outlines are obviously not compatible, and the fact that it accidentally "works" in one environment is no reason for this to work everywhere.

(Sorry to interfere – Frank and I just had a private discussion about this.)

**Comment by justvanrossum** [comment]
In other words: IMO GlyphMath/FontMath (that I assume is responsible for the "working" example is buggy for not complaining about this incompatibility.

ScaleFast is perhaps not designed to catch problems like this (the "Fast" in the name suggests as much), so it can be argued it's actually doing the right thing (the number of points is the same in both masters, it's just the point types that don't match).

**Comment by justvanrossum** [comment]
Looks like RF thinks point 2 is the starting point instead of point 0...
![image](https://user-images.githubusercontent.com/4246121/83523093-e4b6bf00-a4e1-11ea-88a8-f6b98be32153.png)

**Comment by frankrolf** [comment]
Umm … sure. I don’t know if a band-aid like approach like this really is the solution we’re after, but go ahead.
